### PR TITLE
test/relocate.py: skip tests involving patchelf on macOS

### LINKED ISSUE #18450: Relocate tests fails on macOS
When I run `spack test` on macOS, the `relocate.py` tests always fail.

### Steps to reproduce the issue

```console
$ spack test lib/spack/spack/test/relocate.py
...
FAIL lib/spack/spack/test/relocate.py::test_replace_prefix_bin
FAIL lib/spack/spack/test/relocate.py::test_relocate_elf_binaries_absolute_paths
FAIL lib/spack/spack/test/relocate.py::test_relocate_elf_binaries_relative_paths
FAIL lib/spack/spack/test/relocate.py::test_make_elf_binaries_relative
FAIL lib/spack/spack/test/relocate.py::test_relocate_text_bin
...
```

### Error Message

```
E                                  spack.util.executable.ProcessError: Command exited with status 1:
E                                      '/usr/bin/gcc' '-Wl,--disable-new-dtags' '-Wl,-rpath=/usr/lib:/usr/lib64' '/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/pytest-of-Adam/pytest-41/test_replace_prefix_bin0/main.c' '-o' '/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/pytest-of-Adam/pytest-41/test_replace_prefix_bin0/main.x'

lib/spack/spack/util/executable.py:189: ProcessError
-------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------
ld: unknown option: --disable-new-dtags
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
The problem is that these tests assume that `/usr/bin/gcc` is actually GCC. On macOS, `/usr/bin/gcc` is actually a symlink to the Apple Clang compilers, which don't accept the `--disable-new-dtags` flag. We either need to check if `/usr/bin/gcc` is actually GCC, or disable the tests entirely on macOS.

I don't think this error is caught by our CI because we don't have patchelf installed. We should install that in our CI to catch these kind of bugs.

### Information on your system

* **Spack:** 0.15.4-580-42a04f5d6c
* **Python:** 3.7.8
* **Platform:** darwin-catalina-ivybridge

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

@gartung

### PR DESCRIPTION
**Pull request by gartung**

### CODE REVIEW

File: lib/spack/spack/test/relocate.py [hunk 1]
```diff
@@ -132,11 +132,12 @@ def _factory(rpaths, message="Hello world!"):
...
-        opts = [
-            '-Wl,--disable-new-dtags',
+        opts = []
+        if spec.platform != 'darwin':
+            opts = ['-Wl,--disable-new-dtags']
+        opts.append(
             '-Wl,-rpath={0}'.format(rpath_str),
-            str(source), '-o', str(executable)
-        ]
+            str(source), '-o', str(executable))
```
Comment (adamjstewart) [inline lib/spack/spack/test/relocate.py]: Needs to be `opts.extend([...])`, not `opts.append(...)`.

**Comment by gartung** [comment]
try this patch
```
diff --git a/lib/spack/spack/test/relocate.py b/lib/spack/spack/test/relocate.py
index a1a8952c9..99934eff2 100644
--- a/lib/spack/spack/test/relocate.py
+++ b/lib/spack/spack/test/relocate.py
@@ -113,6 +113,10 @@ def _factory(output):
     return _factory
 
 
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 @pytest.fixture()
 def hello_world(tmpdir):
     """Factory fixture that compiles an ELF binary setting its RPATH. Relative
```

**Comment by adamjstewart** [comment]
@gartung it doesn't seem to work when you mark a `fixture`. I used the following patch and it worked:
```diff
diff --git a/lib/spack/spack/test/relocate.py b/lib/spack/spack/test/relocate.py
index a1a8952c97..f9e72ea0d7 100644
--- a/lib/spack/spack/test/relocate.py
+++ b/lib/spack/spack/test/relocate.py
@@ -271,6 +271,10 @@ def test_set_elf_rpaths_warning(mock_patchelf):
 
 
 @pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 def test_replace_prefix_bin(hello_world):
     # Compile an "Hello world!" executable and set RPATHs
     executable = hello_world(rpaths=['/usr/lib', '/usr/lib64'])
@@ -283,6 +287,10 @@ def test_replace_prefix_bin(hello_world):
 
 
 @pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 def test_relocate_elf_binaries_absolute_paths(
         hello_world, copy_binary, tmpdir
 ):
@@ -307,6 +315,10 @@ def test_relocate_elf_binaries_absolute_paths(
 
 
 @pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 def test_relocate_elf_binaries_relative_paths(hello_world, copy_binary):
     # Create an executable, set some RPATHs, copy it to another location
     orig_binary = hello_world(rpaths=['lib', 'lib64', '/opt/local/lib'])
@@ -327,6 +339,10 @@ def test_relocate_elf_binaries_relative_paths(hello_world, copy_binary):
 
 
 @pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 def test_make_elf_binaries_relative(hello_world, copy_binary, tmpdir):
     orig_binary = hello_world(rpaths=[
         str(tmpdir.mkdir('lib')), str(tmpdir.mkdir('lib64')), '/opt/local/lib'
@@ -350,6 +366,10 @@ def test_raise_if_not_relocatable(monkeypatch):
 
 
 @pytest.mark.requires_executables('patchelf', 'strings', 'file', 'gcc')
+@pytest.mark.skipif(
+    platform.system().lower() != 'linux',
+    reason='implementation for MacOS still missing'
+)
 def test_relocate_text_bin(hello_world, copy_binary, tmpdir):
     orig_binary = hello_world(rpaths=[
         str(tmpdir.mkdir('lib')), str(tmpdir.mkdir('lib64')), '/opt/local/lib'
```

**Comment by gartung** [comment]
OK. Updated the PR.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**